keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md

```diff
@@ -0,0 +1,232 @@
+triggering the build constraints on the files in these packages.
+
+Some small adjustments may be necessary to the code base to ensure that the
+other packages can build without depending on these packages.
+
+A prototype is available in [kubernetes/kubernetes#80353](https://github.com/kubernetes/kubernetes/pull/80353).
+
+To ensure that this continues to function we will add CI building in this mode,
+and CI running end to end tests against it (see the test plan).
+
+### User Stories [optional]
+
+#### Story 1
+
+As an out of tree cloud provider implementer, I want to develop and test against
+Kubernetes without the in tree providers.
+
+Kubernetes out of tree cloud provider developers will be able to build Kubernetes
+in this mode and build & test their cloud-controller-manager implementations and
+associated tooling against this build in preparation for the actual hard removal
+of the in-tree providers.
+
+#### Story 2
+
+As a [kind](https://github.com/kubernetes-sigs/kind) developer / user I want to 
+use Kubernetes binaries without cloud providers for local clusters.
+
+Developers and users will be able to build local clusters leveraging this mode
+to not pay for cloud providers they are unable to use.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This is implemented using a synthetic `nolegacyproviders` tag in go build
+constraints on the relevant sources. If `GOFLAGS=-tags=nolegacyproviders` is set
+then the legacy cloud provider pacakges will be excluded from the build.
+
+In order to make this work the following additional changes are made:
+
+- Packages that we fully exclude (the legacy provider packages) _must_ contain
+a `doc.go` file or any other file that does NOT contain any code or build 
+constraints. Go will not allow "building" a package without any files passing
+the constraints, however it will happily build a package with no actual methods
+/ variables / ...
+
+- A few locations in the code do not properly use the cloud provider interface,
+some of these must be updated with both a "with provider" version and a
+"without provider" version broken out of the existing code. In particular this 
+includes the in-tree volumes until CSI migration is standard, and the GCE IPAM
+logic in kube-controller-manager.
+
+In particular this adds tags / constraints to:
+- `staging/src/k8s.io/legacy-cloud-providers/*` (constraints on all the providers)
+- `pkg/cloudprovider` (constraints on importing the providers)
+- `pkg/volumes/*`, `cmd/kubelet` (versions with and without the imported providers for in-tree volumes)
+- `pkg/controller/nodeipam`, `cmd/kube-apiserver`, `cmd/kube-controller-manager` (with and without GCE IPAM)
+
+`test/*` is punted to a future follow up, and credential providers are punted
+to [the external credential provider KEP](https://github.com/kubernetes/enhancements/pull/1137).
+
+### Risks and Mitigations
+
+This is only developer facing and shouldn't be a source of risk.
```

Comment:
I think the biggest risk right now is that developers may add cloud provider code outside of known/controlled locations. While this may break things, doing that should break this build as it will pull in not allowed packages. That will allow us to know and control such additions.

Reply:
doing so would break this special build, but should not break the normal build.

this is a risk to the goal itself, but not really a risk to the project, I'm not sure which this section is meant to cover but assumed the latter (possibly incorrectly)